File: advent_of_code/2017/4/4.py
#!/usr/bin/env python3

input = open('input').read()

c1 = c2 = 0
for passwd in input.split('\n'):
    if passwd == '':
        continue
    parts = passwd.split(' ')
    parts_s = [str(sorted(word)) for word in parts]
    if len(parts) == len(set(parts)):
        c1 += 1
    if len(parts_s) == len(set(parts_s)):
        c2 += 1

print(c1)
print(c2)

File: advent_of_code/2022/4/4.hs
#!/usr/bin/env runhaskell

import Data.Char
import Data.List
import Data.List.Split
import Text.Printf


trim :: String -> String
trim = dropWhileEnd isSpace . dropWhile isSpace


parseInp :: String -> ((Integer, Integer), (Integer, Integer))
parseInp s = ((s11, s12), (s21, s22))
  where
    [[s11, s12], [s21, s22]] = (map read . splitOn "-") <$> (splitOn "," s)


checkFullOverlap :: ((Integer, Integer), (Integer, Integer)) -> Bool
checkFullOverlap ((a1, a2), (b1, b2)) = (a1 <= b1 && b2 <= a2) || (b1 <= a1 && a2 <= b2)


checkOverlap :: ((Integer, Integer), (Integer, Integer)) -> Bool
checkOverlap ((a1, a2), (b1, b2)) = (a1 <= b1 && b1 <= a2) || (b1 <= a1 && a1 <= b2)


main = do
    content <- readFile "input"
    -- content <- readFile "input_test"
    let lc = lines . trim $ content
    let lp = map parseInp lc

    let s1 = sum $ map (fromEnum . checkFullOverlap) lp
    let s2 = sum $ map (fromEnum . checkOverlap) lp

    printf "%d\n%d\n" s1 s2
